packages/runtime/src/runtime.ts

```diff
@@ -736,141 +595,139 @@ export class Runtime {
+   * - Runs TEAL code if associated with transaction
+   * - Executes the transaction on ctx.state
+   * @param txnParams Transaction Parameters
    */
   /* eslint-disable sonarjs/cognitive-complexity */
-  executeTx (txnParams: ExecParams | ExecParams[]): void {
-    const [tx, gtxs] = this.createTxnContext(txnParams); // get current txn and txn group (as encoded obj)
-    // validate first and last rounds
-    this.validateTxRound(gtxs);
-
-    // initialize context before each execution
-    this.ctx = {
-      state: cloneDeep(this.store), // state is a deep copy of store
-      tx: tx,
-      gtxs: gtxs,
-      args: []
-    };
-
-    const txnParameters = Array.isArray(txnParams) ? txnParams : [txnParams];
-    // Run TEAL program associated with each transaction without interacting with store.
-    for (const txnParam of txnParameters) {
+  processTransactions (txnParams: ExecParams[]): void {
+    let clearErrorBool = false;
+    txnParams.forEach((txnParam, idx) => {
+      this.ctxOp.deductFee(txnParam.fromAccount.addr, idx);
       this.assertAmbiguousTxnParams(txnParam);
+
       if (txnParam.sign === SignType.LogicSignature) {
         this.validateLsigAndRun(txnParam);
       }
 
       // https://developer.algorand.org/docs/features/asc1/stateful/#the-lifecycle-of-a-stateful-smart-contract
       switch (txnParam.type) {
-        case TransactionType.CallNoOpSSC:
-        case TransactionType.CloseSSC:
-        case TransactionType.DeleteSSC: {
+        case TransactionType.TransferAlgo: {
+          this.ctxOp.transferAlgo(txnParam);
+          break;
+        }
+        case TransactionType.TransferAsset: {
+          this.ctxOp.transferAsset(txnParam);
+          break;
+        }
+        case TransactionType.CallNoOpSSC: {
           const appParams = this.getApp(txnParam.appId);
           this.run(appParams["approval-program"], ExecutionMode.STATEFUL);
           break;
         }
+        case TransactionType.CloseSSC: {
+          const appParams = this.getApp(txnParam.appId);
+          this.run(appParams["approval-program"], ExecutionMode.STATEFUL);
+          this.ctxOp.closeApp(txnParam.fromAccount.addr, txnParam.appId);
+          break;
+        }
         case TransactionType.ClearSSC: {
           const appParams = this.assertAppDefined(txnParam.appId, this.getApp(txnParam.appId));
-          const fromAccount = this.getAccount(txnParam.fromAccount.addr);
           try {
             this.run(appParams["clear-state-program"], ExecutionMode.STATEFUL);
           } catch (error) {
             // if transaction type is Clear Call, remove the app first before throwing error (rejecting tx)
             // https://developer.algorand.org/docs/features/asc1/stateful/#the-lifecycle-of-a-stateful-smart-contract
-            fromAccount.closeApp(txnParam.appId); // remove app from local state
-            throw error;
+            if (error instanceof RuntimeError && error.number === 1007) {
```

Comment:
Please add a comment regarding error no. `1007`.

Reply:
ok, this is because we only closeout if error is rejected by logic, other than that we should throw error,
we were missing this earlier. will add in code also

Reply:
ok

Reply:
just to confirm something, what about the errors thrown during execution other than `rejected_by_logic`? Let's say `txn Accounts` throws `index_out_of_bound` error (which happens during execution of `clear.teal`). We clear the state in these cases as well, Right?

Reply:
No we don't.

Reply:
I am not sure, that's what i wanted to confirm here. Basically when we execute TEAL on blockchain (with index out of bound error, or something similar), does that constitute a tx `rejected_by_logic` there or not

Reply:
https://developer.algorand.org/docs/features/asc1/teal/#getting-transaction-properties

Reply:
updated as per network behaviour